File: app/src/main/java/deltazero/amarok/apphider/RootAppHider.java
package deltazero.amarok.apphider;

import android.content.Context;

import com.topjohnwu.superuser.Shell;

import java.util.Set;

import deltazero.amarok.R;

public class RootAppHider extends BaseAppHider {

    public RootAppHider(Context context) {
        super(context);
    }

    @Override
    public void hide(Set<String> pkgNames, boolean disableOnly) {
        for (String p : pkgNames) {
            String escaped = "'" + p.replace("'", "'\\''") + "'";
            if (disableOnly) {
                // Only disable, skip hide step
                Shell.cmd(String.format("pm disable %s", escaped)).submit();
            } else {
                // Normal behavior: both disable and hide
                Shell.cmd(String.format("pm disable %s & pm hide %s", escaped, escaped)).submit();
            }
        }
    }

    @Override
    public void unhide(Set<String> pkgNames) {
        for (String p : pkgNames) {
            String escaped = "'" + p.replace("'", "'\\''") + "'";
            Shell.cmd(String.format("pm unhide %s & pm enable %s", escaped, escaped)).submit();
        }
    }

    @Override
    public void tryToActivate(ActivationCallbackListener activationCallbackListener) {
        Shell.getShell(shell -> {
            if (shell.isRoot()) {
                activationCallbackListener.onActivateCallback(RootAppHider.class, true, 0);
            } else {
                activationCallbackListener.onActivateCallback(RootAppHider.class, false, R.string.root_not_ava);
            }
        });
    }

    @Override
    public String getName() {
        return "Root";
    }

}

File: app/src/main/java/deltazero/amarok/apphider/RootAppHider.kt
package deltazero.amarok.apphider

import com.topjohnwu.superuser.Shell
import deltazero.amarok.R
import deltazero.amarok.core.ActivationResult
import deltazero.amarok.core.Hider
import deltazero.amarok.utils.await
import deltazero.amarok.utils.awaitShell

class RootAppHider(private val options: AppHiderOptions) : AppHider {
  override val mode = AppHiderMode.ROOT
  override val name = "Root"

  override suspend fun activate(): ActivationResult {
    val shell = awaitShell()
    return if (shell.isRoot) {
      ActivationResult(success = true, msgResId = 0)
    } else {
      ActivationResult(success = false, msgResId = R.string.root_not_ava)
    }
  }

  override suspend fun process(pkgNames: Set<String>, action: Hider.Action) {
    for (p in pkgNames) {
      when (action) {
        Hider.Action.HIDE -> {
          if (options.disableOnly) {
            Shell.cmd("pm disable $p").await()
          } else {
            Shell.cmd("pm disable $p & pm hide $p").await()
          }
        }
        Hider.Action.UNHIDE -> {
          Shell.cmd("pm unhide $p & pm enable $p").await()
        }
      }
    }
  }
}
